### PR TITLE
feat: enhancement of device mode transformation

### PR DESCRIPTION
**Pull request by MoumitaM**
## PR Description

- Handled new configuration from sourceConfig
  - remove areTranformationsConnected
  - add instead:
    - shouldApplyDeviceModeTransformation: boolean
    - propagateEventsUntransformedOnError: boolean
- Added retry logic based on new config 

## Notion ticket

Ticket link

## Screenshots

Please add screenshots for any new features or UI bug fixes for the following browsers -

- Chrome
  >
- Firefox
  >
- Safari
  >

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by saikumarrs — COMMENTED** [review]
The overall logic is convoluted. Please refactor.

**Review by bardisg — APPROVED** [review]

**Comment by bardisg** [comment]
@MoumitaM lets wait for the call later to see if anything has to change before merging

File: src/features/core/deviceModeTransformation/transformationHandler.js [hunk 1]
```diff
@@ -61,67 +62,69 @@ class TransformationsHandler {
...
               let { response } = xhr;
 
-              if (status === 200) {
-                if (response && typeof response === 'string') {
-                  response = JSON.parse(response);
-                } else {
-                  reject(`[Transformation]:: Transformation failed. Invalid response from server.`);
+              switch (status) {
+                case 200: {
+                  if (response && typeof response === 'string') {
```
Comment (saikumarrs) [inline src/features/core/deviceModeTransformation/transformationHandler.js:67]: What if the response is undefined or not a string?

File: src/features/core/deviceModeTransformation/transformationHandler.js [hunk 2]
```diff
@@ -61,67 +62,69 @@ class TransformationsHandler {
...
+                          .then(resolve)
+                          .catch(reject),
+                      RETRY_INTERVAL * backoffFactor ** (this.retryAttempt - newRetryAttempt),
+                    );
+                  } else {
+                    // Even after all the retries event transformation
+                    // is not successful, return the response with a flag retryExhausted
+                    resolve({
+                      status,
+                      errorMessage: 'Retry exhausted',
```
Comment (saikumarrs) [inline src/features/core/deviceModeTransformation/transformationHandler.js:123]:
```suggestion
                      errorMessage: 'Retries exhausted',
```

File: src/features/core/deviceModeTransformation/transformationHandler.js [hunk 3]
```diff
@@ -166,7 +169,7 @@ class TransformationsHandler {
         }
         this.isTransformationProcessing = false;
         // send null as response in case of error or retry fail
```
Comment (saikumarrs) [inline src/features/core/deviceModeTransformation/transformationHandler.js:171]: Please update this comment.

File: src/utils/ObjectUtils.js [hunk 4]
```diff
@@ -54,11 +54,15 @@ const stringifyWithoutCircular = (obj, excludeNull) =>
 const isInstanceOfEvent = (value) =>
   typeof value === 'object' && value !== null && 'target' in value;
 
+const isNonEmptyObject = (value) =>
+  typeof value === 'object' && value !== null && Object.keys(value).length > 0;
```
[the remark this answers is not in the document]
Comment (bardisg) [reply src/utils/ObjectUtils.js:58]: why we do not use isObjectLiteralAndNotNull existing method?

File: src/core/analytics.js [hunk 5]
```diff
@@ -404,57 +408,99 @@ class Analytics {
...
+   * @param {Array} destWithTransformation   List of device mode destination that has transformation connected
+   * @param {Object} rudderElement           Rudder event
+   * @param {string} methodName              Type of event page/track/identify etc.
+   */
   sendTransformedDataToDestination(destWithTransformation, rudderElement, methodName) {
     try {
       // convert integrations object to server identified names, kind of hack now!
       transformToServerNames(rudderElement.message.integrations);
-
+      const destinationIds = destWithTransformation.map((e) => e.destinationId);
```
Comment (bardisg) [inline src/core/analytics.js:422]: does destWithTransformation also take into acocunt event filtering for the destination?

File: src/utils/ObjectUtils.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/utils/ObjectUtils.js:58]: @bardisg `isObjectLiteralAndNotNull` will return true for an empty object.

[the remark this answers is not in the document]
Comment (saikumarrs) [reply src/utils/ObjectUtils.js:58]: I think he meant to use it in conjunction with the length check.

[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/utils/ObjectUtils.js:58]: Got it.

File: src/core/analytics.js (hunk 5 above)
Comment (MoumitaM) [reply src/core/analytics.js:422]: Yes

File: src/core/analytics.js [hunk 6]
```diff
@@ -404,57 +408,99 @@ class Analytics {
...
+                    // send only if the event is not null/undefined/empty object
+                    if (isNonEmptyObject(tEvent.event)) {
+                      this.sendDataToDestination(intg, { message: tEvent.event }, methodName);
+                    }
+                  });
+                  break;
+                }
+                // Transformation server access denied
+                case 404: {
+                  // send untransformed event to destination
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/core/analytics.js:476]: This is not a failure case. This is for non-customers who accessed DMT somehow.

[the remark this answers is not in the document]
Comment (saikumarrs) [reply src/core/analytics.js:476]: Yes but we should be able to identify those scenarios.

File: src/features/core/deviceModeTransformation/transformationHandler.js (hunk 1 above)
Comment (MoumitaM) [reply src/features/core/deviceModeTransformation/transformationHandler.js:67]: That is the contract we have with the server. 200 will have some response in correct format

Comment (saikumarrs) [reply src/features/core/deviceModeTransformation/transformationHandler.js:67]: In that case, there is no need for the 'if' check.

File: src/utils/ObjectUtils.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/utils/ObjectUtils.js]: Adressed

File: src/core/analytics.js [hunk 7]
```diff
@@ -404,57 +408,99 @@ class Analytics {
...
                     (e) => e.id === intg.destinationId,
                   );
-                  if (!destTransformedResult) {
-                    logger.error(
-                      `[DMT]::Transformed data for destination "${intg.name}" was not sent from the server`,
-                    );
-                    return;
-                  }
-
+                  const eventsToSent = [];
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/core/analytics.js]: Adressed

File: src/features/core/deviceModeTransformation/transformationHandler.js (hunk 1 above)
Comment (MoumitaM) [reply src/features/core/deviceModeTransformation/transformationHandler.js]: Adressed

File: src/features/core/deviceModeTransformation/transformationHandler.js (hunk 3 above)
Comment (MoumitaM) [reply src/features/core/deviceModeTransformation/transformationHandler.js]: Adressed

File: src/integrations/AdobeAnalytics/browser.js [hunk 8]
```diff
@@ -31,8 +31,10 @@ class AdobeAnalytics {
     this.proxyHeartbeatUrl = config.proxyHeartbeatUrl;
     this.pageName = '';
     this.name = NAME;
-    this.areTransformationsConnected =
-      destinationInfo && destinationInfo.areTransformationsConnected;
+    this.shouldApplyDeviceModeTransformation =
+      destinationInfo && destinationInfo.shouldApplyDeviceModeTransformation;
+    this.propagateEventsUntransformedOnError =
+      destinationInfo && destinationInfo.propagateEventsUntransformedOnError;
     this.destinationId = destinationInfo && destinationInfo.destinationId;
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/integrations/AdobeAnalytics/browser.js]: Adressed

File: src/core/analytics.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/core/analytics.js:475]: It is a status code that defines the customer doesn't have access to DMT(SDK doesn't need to know the reason ) and the SDK should send the original event to the destination.
Do we really want to warn our customers about something they are not responsible for?

[the remark this answers is not in the document]
Comment (MoumitaM) [reply src/core/analytics.js:476]: Adressed

File: src/core/analytics.js [hunk 9]
```diff
@@ -474,6 +472,7 @@ class Analytics {
                 }
                 // Transformation server access denied
                 case 404: {
+                  logger.warn(`System failure`);
```
Comment (saikumarrs) [inline src/core/analytics.js:475]: The warning message could be more explicit.
```suggestion
                  logger.warn(`[DMT]:: Transformation server access is denied. The configuration data seems to be out of sync. Sending untransformed event to the destination.`);
```

File: src/core/analytics.js [hunk 10]
```diff
@@ -434,37 +434,35 @@ class Analytics {
...
-                    // and propagateEventsUntransformedOnError is not true for that destination
-                    else {
-                      let errMessage = `[DMT]::Event transformation unsuccessful for destination "${intg.name}". Status: "${tEvent.status}". Dropping the event.`;
-                      if (tEvent.status === 410) {
-                        errMessage = `[DMT]::Event transformation unsuccessful for destination "${intg.name}" as the transformation is not available. Status: "${tEvent.status}". Dropping the event.`;
+
+                      let action = 'Dropping the event';
+                      let logMethod = logger.error;
+                      if (intg.propagateEventsUntransformedOnError === true) {
+                        action = 'Sending untransformed event';
```
Comment (saikumarrs) [inline src/core/analytics.js:455]:
```suggestion
                        action = 'Sending untransformed event to the destination';
```

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/core/analytics.js (hunk 9 above)
Comment (MoumitaM) [reply src/core/analytics.js:475]: chore: error message update

**Review by saikumarrs — APPROVED** [review]

**Review by bardisg — DISMISSED** [review]

**Review by bardisg — APPROVED** [review]

**Review by bardisg — APPROVED** [review]

**Pull request merged**